soc/xtensa/intel_adsp/common/include/intel_adsp_ipc.h

```diff
@@ -174,4 +178,38 @@ bool intel_adsp_ipc_send_message_sync(const struct device *dev,
 void intel_adsp_ipc_send_message_emergency(const struct device *dev, uint32_t data,
 					   uint32_t ext_data);
 
-#endif /* ZEPHYR_INCLUDE_INTEL_ADSP_IPC_H */
+#ifdef CONFIG_PM_DEVICE
+
+typedef int (*intel_adsp_ipc_resume_handler_t)(const struct device *dev, void *arg);
+
+typedef int (*intel_adsp_ipc_suspend_handler_t)(const struct device *dev, void *arg);
+
+/**
+ * @brief Registers wake callback handler used to wake device from sleep.
+ *
+ * @param dev IPC device.
+ * @param fn Callback function.
+ * @param arg Value to pass as the "arg" parameter to the function.
+ */
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg);
+
+/**
+ * @brief Registers suspend callback handler used to put device to sleep.
+ *
+ * @param dev IPC device.
+ * @param fn Callback function.
+ * @param arg Value to pass as the "arg" parameter to the function.
+ */
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg);
```

Comment:
This is not a blocker for me but I would avoid this two way dependency. The suspend callback is doing almost nothing and second one for resume is reinitializing edf. This operation can be done before FW Redy notification in notify_pm_state_exit.

Reply:
Ok, so there are few aspects here I would like to refer.

**Aspect 1 - proper code isolation and syntax**
We must have this two way dependency for now as the IPC driver code is split between Zephyr and SOF.  So, in any case we need to invoke IPC functions on the SOF side (like setting mask, initializing edf etc).
Now, you are suggesting that I should expose IPC driver private functions straight to the notify_pm_state_exit function.
I do not agree with this, since device initialization is a "private implementation" of the driver that should not be exposed outside. If we start to call private driver functions from everywhere bypassing interfaces and proper abstractions we would get a spaghetti code in no time.

**Aspect 2 - usefulness of the intel_adsp_ipc_suspend_handler**
* The `pm_policy_state_lock_is_active` lock used in the resume handler works even if power state is changed by force (with `pm_state_force()`)
* Given, that only IPC message is a source of signal for the whole system to enter D3 power state, we should validate consistency of IPC API. By his I mean, that when  `pm_state_force()` is called to enter D3, we must make sure that the `ipc->task_mask & IPC_TASK_POWERDOWN` is set in the SOF API part. Otherwise it is a fatal system condition.

Reply:
This API is really confuse. Which device it is supposed to put to sleep ? The SoC ? If that is the case the whole patch is problematic and should not be done this way. The SoC should only be suspened from the idle thread.

Be careful with the nomemclature, from the Zephyr perspective device means peripherals.

Reply:
> Be careful with the nomemclature, from the Zephyr perspective device means peripherals.

From the Zephyr perspective, "Device" means a entity that implements [Device Driver Model](https://docs.zephyrproject.org/3.1.0/kernel/drivers/index.html), not peripherals. Maybe I forgot to write it with capital letter but in general my nomenclature is very precise. IPC is a Zephyr driver that implements Device Driver Model. This patch puts to sleep IPC Device.

> The SoC should only be suspened from the idle thread.

And it is. Unfortunately IPC driver API is split partially across SOF and Zephyr, that is why some things are very confusing.
This is done this way, because SOF maintains also IPC3 which does not use Zephyr IPC driver. SOF builds a façade on common IPC functions. This Zephyr IPC "API" is used from `sof\src\ipc\ipc-zephyr.c`.
To fully understand this PR, you should also have a look at the content of https://github.com/thesofproject/sof/pull/7325 .
Suspending logic goes like this:
1.  `sof\src\ipc\ipc-zephyr.c: ipc_platform_do_cmd()` function calls cpu_disable_core(PLATFORM_PRIMARY_CORE_ID) function. This sends instruction to Zephyr Power Manager, to enter D3 power state as soon as we enter idle (from idle thread). Feature now works for the primary core only, secondary cores D3 are under review too: https://github.com/zephyrproject-rtos/zephyr/pull/55182 .
2. Zephyr Power Manager during idle thread starts power state change (to PM_STATE_SOFT_OFF what we call D3).
3. Power state change invokes [PM notifiers](https://docs.zephyrproject.org/2.7.0/reference/power_management/index.html#structpm__notifier) callbacks when Power State is entered and exited.
Those are registered functions in `sof\src\platform\intel\ace\platform.c` (See https://github.com/thesofproject/sof/pull/7325/commits/defab6cf0bf1e70748a6b33f2bc4da26c9c4e7e5):
`static void notify_pm_state_entry(enum pm_state state)`
`static void notify_pm_state_exit(enum pm_state state)`
4. As we start to enter D3 state, `static void notify_pm_state_entry(enum pm_state state)` gets called. In this function body we prepare all Devices (and other entities) in the SoC to enter D3 state and do a context save (memory dump).
Function `pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV, PM_DEVICE_ACTION_SUSPEND);` gets called that is a IPC Device PM API registered to [Zephyr Device Power Manager API](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#device-model-with-power-management-support).
5. Calling `pm_device_action_run` on IPC Device, invokes the IPC Device power change handler:
`static int ipc_pm_action(const struct device *dev, enum pm_device_action action)`.
Inside this handler, we do following things:
    * mark IPC Device as busy - it now performs a power state change and should not be used. We set a proper status -  `pm_device_busy_set(dev);`.
    * take IPC Device PM lock - Device may not be asked to do power state change as it currently processes one:
`pm_device_state_lock(dev);`
    * At this point we invoke a callback to SOF IPC4 code as we need to validate the state of IPC subsystem.
    This is done using registered previously (with `intel_adsp_ipc_set_suspend_handler()`) function callback to: `sof\src\ipc\ipc-zephyr.c: ipc_device_suspend_handler()`
    * Inside this handler, we validate whether: IPC message queue is empty and `IPC_TASK_POWERDOWN` flag is set.
    If it is not, something is wrong with our internal messaging system and we terminate FW using assertion with proper error message indicating a reason.
    * If everything is ok, we release PM lock and set device state as not busy anymore. IPC Device power state had been changed to D3. The end.

Similar logic applies when we exit D3 state, just other notifier gets called, other callback to SOF IPC used and it does IPC Device reinitialization after waking up (enabling interrupts cleaning SOF task masks).

Reply:
how are you calling `pm_device_action_run` ? I suspect you should be using device runtime. `pm_device_runtime_get` and `pm_device_runtime_put`   When device runtime is enabled the device is not suspended / resumed when the system and you do it manually, which is basically what you are doing. This will simplify this code a lot since you don't need to worry about ensuring that device is not suspended without you trigger it.